lib/backend/memory/memory.go

```diff
@@ -480,35 +480,15 @@ func (m *Memory) processEvent(event backend.Event) {
 		if treeItem != nil {
 			existingItem = treeItem.(*btreeItem)
 		}
-		switch {
-		case item.Expires.IsZero():
-			// new item is added, it does not expire,
-			if existingItem != nil && existingItem.index >= 0 {
-				// new item replaces the existing item that should be removed
-				// from the heap
-				m.heap.RemoveEl(existingItem)
-			}
-			m.tree.ReplaceOrInsert(item)
-		case !item.Expires.IsZero() && m.Clock().Now().Before(item.Expires):
-			// new item is added, but it has not expired yet
-			if existingItem != nil && existingItem.index >= 0 {
-				m.heap.RemoveEl(existingItem)
-			}
+
+		// new item is added, but it has not expired yet
+		if existingItem != nil && existingItem.index >= 0 {
+			m.heap.RemoveEl(existingItem)
+		}
+		if !item.Expires.IsZero() {
 			m.heap.PushEl(item)
-			m.tree.ReplaceOrInsert(item)
-		case !item.Expires.IsZero() && (m.Clock().Now().After(item.Expires) || m.Clock().Now() == item.Expires):
-			// new expired item has added, remove the existing
-			// item if present
-			if existingItem != nil {
-				// existing item should be removed from the heap
-				if existingItem.index >= 0 {
-					m.heap.RemoveEl(existingItem)
-				}
-				m.tree.Delete(existingItem)
-			}
-		default:
-			// skip adding or updating the item that has expired
 		}
+		m.tree.ReplaceOrInsert(item)
```

Comment:
Is the bug because originally ReplaceOrInsert wouldn't run in a few branches?

Reply:
That, and if the item was expired it would actually remove the old value.  Basically, "put" events for expired items were treated as removals, but that isn't really what we want.  Its better for the cache to be outdated but correct than for it to stray from the backend state.  Until the backend emits an event indicating expiration, the cache shouldn't assume that the value is *really* expired, just because the cache's current view indicates expiry.

Reply:
Got it, thanks for the thoughtful explanation.